keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/TimeWindowConstraintEvaluator.kt

```diff
@@ -0,0 +1,203 @@
+import com.netflix.spinnaker.keel.constraints.ConstraintEvaluator.Companion.getConstraintForEnvironment
+import com.netflix.spinnaker.keel.exceptions.InvalidConstraintException
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import org.springframework.stereotype.Component
+import java.text.ParsePosition
+import java.time.Clock
+import java.time.DayOfWeek
+import java.time.format.DateTimeFormatter
+import java.time.format.TextStyle
+import java.util.Locale
+import java.util.TimeZone
+import java.util.TimeZone.getTimeZone
+
+/**
+ * An environment promotion constraint to gate promotions to time windows.
+ * Multiple windows can be specified.
+ *
+ * Hours and days parameters support a mix of ranges and comma separated values.
+ *
+ * Days can be specified using full or short names according to the default JVM locale.
+ *
+ * A timezone can be set in the constraint and a system-wide default can by set via
+ * the dynamic property "default.time-zone" which defaults to UTC.
+ *
+ * Example env constraint:
+ *
+ * constraints:
+ *  - type: time-window
+ *    windows:
+ *      - days: Monday,Tuesday-Friday
+ *        hours: 11-16
+ *        tz: America/Los_Angeles
+ *      - days: Wed
+ *        hours: 12-14
+ *        tz: America/Los_Angeles
+ */
+@Component
+class TimeWindowConstraintEvaluator(
+  private val clock: Clock,
+  private val dynamicConfigService: DynamicConfigService
+) : ConstraintEvaluator<TimeWindowConstraint> {
+  companion object {
+    val whiteSpace = """\s""".toRegex()
+    val intOnly = """^\d+$""".toRegex()
+    val intRange = """^\d+\-\d+$""".toRegex()
+
+    val fullDayFormatter: DateTimeFormatter = DateTimeFormatter
+      .ofPattern("EEEE", Locale.US)
+    val shortDayFormatter: DateTimeFormatter = DateTimeFormatter
+      .ofPattern("EEE", Locale.US)
+
+    val daysOfWeek = DayOfWeek.values()
+      .map {
+        listOf(
+          it.toString().toLowerCase(),
+          it.getDisplayName(TextStyle.SHORT, Locale.getDefault()).toLowerCase()
+        )
+      }
+      .flatten()
+      .toMutableSet()
+  }
+
+  override val constraintType = TimeWindowConstraint::class.java
+
+  override fun canPromote(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: String
+  ): Boolean {
+    val constraint = getConstraintForEnvironment(deliveryConfig, targetEnvironment, constraintType)
+
+    val tz: TimeZone = if (constraint.tz != null) {
+      getTimeZone(constraint.tz)
+    } else {
+      defaultTimeZone()
+    }
+
+    val now = clock
+      .instant()
+      .atZone(tz.toZoneId())
+
+    constraint.windows.forEach {
+      val hours = parseHours(it.hours, deliveryConfig, targetEnvironment)
+      val hoursMatch = hours.isEmpty() || hours.contains(now.hour)
+
+      val today = now.dayOfWeek
+        .toString()
+        .toLowerCase()
+      val todayShort = now.dayOfWeek
+        .getDisplayName(TextStyle.SHORT, Locale.getDefault())
```

Comment:
Shouldn't this be `Locale.US` for consistency with the formatters up above?

Reply:
I think it's the other way around, they should be Locale.default. Thanks for pointing out the discrepancy.